src/treecontroller/viewconsumable.js

```diff
@@ -0,0 +1,486 @@
+
+			if ( value === false ) {
+				consumables.set( name, true );
+			}
+		}
+	}
+}
+
+/**
+ * Class used for handling consumption of {@link engine.treeView.Element view Elements}.
+ * Element's name and its parts (attributes, classes and styles) can be consumed separately. Consuming an element's name
+ * does not consume its attributes, classes and styles.
+ * To add items for consumption use {@link engine.treeController.ViewConsumable#add add method}.
+ * To test items use {@link engine.treeController.ViewConsumable#test test method}.
+ * To consume items use {@link engine.treeController.ViewConsumable#consume consume method}.
+ * To revert already consumed items use {@link engine.treeController.ViewConsumable#revert revert method}.
+ *
+ *		viewConsumable.add( element, { name: true } ); // Adds element's name as ready to be consumed.
+ *		viewConsumable.test( element, { name: true }  ); // Tests if element's name can be consumed.
+ *		viewConsumable.consume( element, { name: true }  ); // Consume element's name.
+ *		viewConsumable.revert( element, { name: true }  ); // Revert already consumed element's name.
+ *
+ * @memberOf engine.treeController
+ */
+export default class ViewConsumable {
+
+	/**
+	 * Creates new ViewConsumable.
+	 */
+	constructor() {
+		/**
+		* Map of consumable elements.
+		*
+		* @protected
+		* @member {Map.<engine.treeController.ViewElementConsumables>} engine.treeController.ViewConsumable#_consumables
+		*/
+		this._consumables = new Map();
+	}
+
+	/**
+	 * Adds {@link engine.treeView.Element view Element} and its parts as ready to be consumed.
+	 *
+	 *		viewConsumable.add( p, { name: true } ); // Adds element's name to consume.
+	 *		viewConsumable.add( p, { attribute: 'name' } ); // Adds element's attribute.
+	 *		viewConsumable.add( p, { class: 'foobar' } ); // Adds element's class.
+	 *		viewConsumable.add( p, { style: 'color' } ); // Adds element's style
+	 *		viewConsumable.add( p, { attribute: 'name', style: 'color' } ); // Adds attribute and style.
+	 *		viewConsumable.consume( p, { class: [ 'baz', 'bar' ] } ); // Multiple consumables can be provided.
+	 *
+	 * Throws {@link utils.CKEditorError CKEditorError} `viewconsumable-invalid-attribute` when `class` or `style`
+	 * attribute is provided - it should be handled separately by providing actual style/class.
+	 *
+	 *		viewConsumable.add( p, { attribute: 'style' } ); // This call will throw an exception.
+	 *		viewConsumable.add( p, { style: 'color' } ); // This is properly handled style.
+	 *
+	 * @param {engine.treeView.Element} element
+	 * @param {Object} consumables
+	 * @param {Boolean} consumables.name If set to true element's name will be included.
+	 * @param {String|Array} consumables.attribute Attribute name or array of attribute names.
+	 * @param {String|Array} consumables.class Class name or array of class names.
+	 * @param {String|Array} consumables.style Style name or array of style names.
+	 */
+	add( element, consumables ) {
```

Comment:
For all these methods `element` may be not only the element, but also a text node or document fragment (root). For all there there is not `consumables` details, but it should be possible to consume text node or document fragment.